File: src/main/java/io/digitalstate/stix/common/StixBoolean.java
package io.digitalstate.stix.common;

/**
 * Wrapper for boolean that is used in Stix to track if the boolean value was User provided.
 * Custom class is used instead of a Boolean in order to provide a more robust api
 */
public class StixBoolean {

    private boolean stixBooleanValue;
    private boolean isDefinedValue;

    public StixBoolean(boolean stixBooleanValue) {
        this.stixBooleanValue = stixBooleanValue;
        this.isDefinedValue = true;
    }

    public StixBoolean(boolean stixBooleanValue, boolean isDefinedValue) {
        this.stixBooleanValue = stixBooleanValue;
        this.isDefinedValue = isDefinedValue;
    }

    /**
     * Defaults to StixBoolean value to false.  Sets isDefinedValue to false
     */
    public StixBoolean() {
        this.stixBooleanValue = false;
        isDefinedValue = false;
    }

    public StixBoolean(String booleanString){
        this.stixBooleanValue = Boolean.valueOf(booleanString);
        this.isDefinedValue = true;
    }

    public boolean getStixBooleanValue() {
        return stixBooleanValue;
    }

    /**
     * Indicates that the boolean value was explicitly defined, even if the value was false,
     * and the original object's property defaults to false if no value is provided.
     * @return boolean indicating if the value was defined.
     */
    public boolean isdefinedValue() {
        return isDefinedValue;
    }

    @Override
    public String toString() {
        return String.valueOf(getStixBooleanValue());
    }

}

File: src/main/kotlin/com/stephenott/stix/type/StixBoolean.kt
package com.stephenott.stix.type

data class StixBoolean(
    val value: Boolean = false,
    val isDefinedValue: Boolean = false
) {

    /**
     * Defaults to StixBoolean value to false.  Sets isDefinedValue to false.
     * Essentially if the no arg constructor was used, we assume that its a default boolean value rather than explicitly being provided.
     * This mainly has implications for JSON processing.
     */
    constructor() : this(false, false)

    constructor(value: Boolean) : this(value, true)

    constructor(stixBoolean: StixBoolean) : this(stixBoolean.value, stixBoolean.isDefinedValue)

    companion object {
        fun parse(booleanString: String): StixBoolean {
            return StixBoolean(booleanString.toBoolean(), true)
        }
    }

    override fun toString(): String {
        return this.value.toString()
    }

}